Atomic/syscontainers.py

```diff
@@ -922,34 +919,44 @@ def _pull_dockertar_layers(self, repo, imagebranch, temp_dir, input_layers):
             layers_to_import[layers_map[k]] = v
         SystemContainers._import_layers_into_ostree(repo, imagebranch, manifest, layers_to_import)
 
-    @staticmethod
-    def _generate_validation_manifest(layer,tar,root=os.path.join(ATOMIC_VAR, "gomtree-manifests")):
-        """
-        Creates a gomtree validation manifest using a tar archive associated with
-        it's layer. Defaults to placing the manifest in a folder in the ATOMIC_VAR folder.
-        :param layer: layer digest
-        :param tar: may or may not be gzip compressed
-        :return: None
-        """
-        if not os.path.exists(root):
-            os.makedirs(root)
-        manifestname = os.path.join(root, "%s.mtree" % layer)
-        if os.path.isfile(manifestname):
-            util.write_out("validation manifest for layer %s already created" % layer)
-        else:
-            gomtree = util.generate_validation_manifest(img_tar=tar,keywords="type,uid,gid,mode,size,sha256digest")
-            if gomtree.return_code != 0:
-                decompressed = tar.replace(".tar",".tar.gz")
-                os.rename(tar,decompressed)
-                with gzip.open(decompressed, "rb") as g:
-                    _, tmptar = tempfile.mkstemp()
-                    with open(tmptar, "wb") as t:
-                        t.write(g.read())
-                        gomtree = util.generate_validation_manifest(img_tar=tmptar,keywords="type,uid,gid,mode,size,sha256digest")
-                    os.remove(tmptar)
-                os.rename(decompressed, tar.replace(".tar.gz",".tar"))
-            if gomtree.return_code == 0:
-                with open(manifestname, "w", 0) as m:
-                    m.write(gomtree.stdout)
-            else:
-                util.write_out(gomtree.stderr)
+    def validate_layer(self, layer):
+        ret = []
+        layer = layer.replace("sha256:", "")
+        repo = self._get_ostree_repo()
+        if not repo:
+            return ret
+
+        def validate_ostree_file(csum):
+            _, inputfile, file_info, xattrs = repo.load_file(csum)
+            # images are imported from layer tarballs, without any xattr.  Don't use xattr to compute
+            # the OSTree object checksum.
+            xattrs = GLib.Variant("a(ayay)", [])
+            _, checksum_v = OSTree.checksum_file_from_input(file_info, xattrs, inputfile, OSTree.ObjectType.FILE)
+            return OSTree.checksum_from_bytes(checksum_v)
+
+        def traverse(it):
```

Comment:
I think this code is all OK, but we should consider pushing it down into OSTree at some point.